proposals/js-string-builtins/Overview.md

```diff
@@ -163,6 +163,20 @@ JS Strings are semantically a sequence of 16-bit code units (referred to as char
 
 There is however, the Encoding API for `TextEncoder`/`TextDecoder` which can be used for UTF-8 support. However, this is technically a separate spec from JS and may not be available on all JS engines (in practice it's available widely). This proposal exposes UTF-8 data conversions using this API under separate `wasm:text-encoder` `wasm:text-decoder` interfaces which are available when the host implements these interfaces.
 
+## String constants
+
+String constants may be defined in JS and made available to wasm through a variety of means.
+
+The simplest way is to have a module import each string as an immutable global. This can work for small amounts of strings, but has a high cost for when the number of string constants is very large.
+
+This proposal adds an extension to JS-API compile routine to support optimized 'imported string constants' to address this use-case.
+
+The `WebAssemblyCompileOptions` is extended with a `boolean importedStringConstants` flag. When this is set, the module may define imports of the form `(import "'" "%stringConstant%"" (global externref))`, and the JS-API will use the provided `%stringConstant%` import field name to be the value of the global. This allows for any UTF-8 string to be imported with minimal overhead.
+
+The string namespace is chosen to be the single quote ASCII character `'`. We may revise this to be a longer name before this proposal is finalized.
+
+All imports that reference this namespace must be globals of type immutable externref. If they are not, an eager compile error is emitted.
```

Comment:
Allowing mutable globals as well would allow imports of invalid UTF-16 to use replacement characters that could then be automatically patched up in the start function. That seems better than having to fall back to an out-of-band import mechanism for invalid strings.

Reply:
Allowing mutable globals has two major drawbacks:
(1) The reason we use an imported global for each string constant in the first place is because we want to `global.get` them from other global initializers, which is only valid for immutable globals. So allowing mutable globals would defeat the primary use case.
(2) Globals that are both magically-populated and mutable take away an engine's ability to initialize them lazily, because they could have been manually mutated in the meantime. While I'm not sure how relevant the ability to initialize globals lazily is in practice, it's nice to have the option.

In addition, I'm not convinced by the described benefit. If you want the start function to populate a mutable global with an invalid-utf8 string constant, just define a non-imported mutable global for that. If you call that an "out of band" mechanism and want something that's closer to how valid-utf8 string constants are handled, just import it (as immutable) with a different module name, and provide the value via embedding JavaScript.

Reply:
I agree with Jakob, immutable is nicer for engines to optimize (one opt is to recognize that these imported globals will always have a string even though they're externref, and avoid some casts), and it also seems like you could define your own mutable globals as a workaround if you need them.